### ISSUE TITLE
`Kokkos::<exec space>::in_parallel` signatures are not consistent across execution spaces

### ISSUE THREAD

**Issue by romintomasetti** [kokkos/kokkos#6032]
**Describe the bug**

The signature of `Kokkos::<execution space>::in_parallel` is not consistent across all execution spaces.

Looking at the below code excerpt:

```c++
Kokkos::Serial { inline static int in_parallel(); }

Kokkos::OpenMP { static bool in_parallel(OpenMP const& = OpenMP()); }

Kokkos::Cuda { KOKKOS_INLINE_FUNCTION static int in_parallel(); }
```

it seems there is a lack of consistency. According to
- #457

it should have the following signature
```c++
static bool in_parallel( ExecutionSpace const& = ExecutionSpace() );
```

I think, as `in_parallel` will be used more often (see for instance trilinos/Trilinos/pull/11679), it is important to keep it consistent.

I would also like to note here that the signature should have `KOKKOS_INLINE_FUNCTION` (or at least `KOKKOS_FUNCTION`), because one should be able to write the following code:
```c++
using execution_space = Kokkos::Serial; // or Kokkos::OpenMP

...

... KOKKOS_LAMBDA(...) { return execution_space::in_parallel();} ...
```

that for now will not compile if warnings are treated as errors, compiling for Cuda (it will stop on `calling a __host__ function("Kokkos::OpenMP::in_parallel(const Kokkos::OpenMP&)") from a __host__ __device__ function(...) is not allowed`).

**Comment by romintomasetti** [comment]
@maartenarnst

**Comment by masterleinad** [comment]
Note that https://kokkos.github.io/kokkos-core-wiki/API/core/execution_spaces.html#synopsis says the interface should be conforming to
```
    bool in_parallel() const;
```
(which `Serial`, `OpenMP`, and `Cuda` are except for the `KOKKOS_FUNCTION` annotation)

**Comment by crtrott** [comment]
This function is actually kind of problematic and so is the use case pointed to above. Essentially right now "in_parallel()" is a mix of

- "I am a thread which is currently participating in executing a Kokkos parallel region"
- "Someone is currently executing a parallel region"
- "I am a thread which runs on the device, but I don't actually know whether I am part of a parallel region"

So getting no does not mean: there is no `parallel_for` of that execution space running. And getting yes does not reliably mean: the calling thread participates in a `parallel_for`.

We need to come up with proper semantics and reimplement something which is actually implementable.

**Comment by ajpowelsnl** [comment]
@crtrott, @masterleinad - should this be an item for Wed. meeting discussion?   Which Wednesday, is another question, but the issue is not getting any younger...

**Comment by masterleinad** [comment]
> @crtrott, @masterleinad - should this be an item for Wed. meeting discussion? Which Wednesday, is another question, but the issue is not getting any younger...

We just need to prioritize and assign. The question is also if this is actually used and if we care to improve or if `KOKKOS_IF_ON_HOST/DEVICE` is sufficient.

**Comment by ajpowelsnl** [comment]
> > @crtrott, @masterleinad - should this be an item for Wed. meeting discussion? Which Wednesday, is another question, but the issue is not getting any younger...
> 
> We just need to prioritize and assign. The question is also if this is actually used and if we care to improve or if `KOKKOS_IF_ON_HOST/DEVICE` is sufficient.

OK, so a yes / no vote on Wed.

**Comment by ajpowelsnl** [comment]
- **General questions, comments**:
- Does the current instance do anything?
  - Original idea: at the call site, determine if you are in a parallel region; am I in a `parallel_for`?
  - Option: thread-local variables that have "state-ness"?
  - If Option, then what does that mean for Serial 
- Are others using this code?
- In documentation on `ExecutionSpace`
- Is it a static function, or a member function, or a const member function?
- What does `in_parallel` mean when you have different backend instances
- Similar to `CUDA_ARCH` , which tells you if you're inside of parallel code
- Bottom line, we don't yet know what the semantics should be
- Should return a `bool` and not an `int`
- Issue #457 is no longer relevant to this discussion
- Option: deprecate this code if it's unused
- **TODO**:
- grep through Trilinos (exluding kokkos dir) to find out if it's used
- one hit Tpetra, two in Stokhos;
- In Stokhos, used in parallel to see if they can call `parallel_for`, using our `impl::parallel_for`
-  Similar in Tpetra
- **Decision (tentative)**:
- Consider deprecating because not widely used, and implementation path is not clear (only alternative thread-local variable)
- Ask user base in Slack if they actually want it, and if they want it, what semantics would you like?

**Comment by ajpowelsnl** [comment]
@masterleinad - would you please ask `#general` if they are interested in this consistency across `ExecutionSpace`, and if so, how would they like the semantics to look?

**Comment by masterleinad** [comment]
The only place this is used in `Trilinos` is the same as in `Kokkos` (`Sacado`/`Stokhos` specializations of `Views`). I think we should just deprecate and see if someone complains.
